File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/mad/Java/C++/s685734377_commented.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <sstream>
#include <string>

using namespace std;

struct FastScanner {
    queue<string> tokens;
    FastScanner(istream& in) {
        string line;
        while (getline(in, line)) {
            istringstream iss(line);
            string token;
            while (iss >> token) {
                tokens.push(token);
            }
        }
    }
    string next() {
        if (tokens.empty()) throw runtime_error("No more tokens");
        string token = tokens.front();
        tokens.pop();
        return token;
    }
    int nextInt() {
        string token = next();
        return stoi(token);
    }
};

class TaskB {
public:
    void solve(int testNumber, FastScanner& in, ostream& out) {
        int n = in.nextInt();
        vector<vector<char>> a(n, vector<char>(n));
        for (int i = 0; i < n; ++i) {
            string s = in.next();
            for (int j = 渴:0; j < n; ++j) {
                a[i][j] = s[j];
            }
        }

        const long long P = 37;
        vector<long long> powP(n + 1);
        powP[0] = 1;
        for (int i = 1; i <= n; ++i) {
            powP[i] = powP[i - 1] * P;
        }

        vector<vector<long long>> hashRow(n, vector<long long>(n, 0));
        vector<vector<long long>> hashCol(n, vector<long long>(n, 0));

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                for (int k = 0; k < n; ++k) {
                    int mod = (j - k + n) % n;
                    hashRow[i][k] += a[i][j] * powP[mod];
                    mod = (i - k + n) % n;
                    hashCol[j][k] += a[i][j] * powP[mod];
                }
            }
        }

        int ans = 0;
        for (int A = 0; A < n; ++A) {
            for (int B = 0; B < n; ++B) {
                bool ok = true;
                for (int i = 0; i < n; ++i) {
                    int row = (i + A) % n;
                    int col = (i + B) % n;
                    if (hashRow[row][B] != hashCol[col][A]) {
                        ok = false;
                        break;
                    }
                }
                if (ok) ans++;
            }
        }

        out << ans << endl;
    }
};

int main() {
    FastScanner in(cin);
    TaskB solver;
    solver.solve(1, in, cout);
    return 0;
}
// End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/mad/Java/Python/s685734377_commented.py
import sys

class FastScanner:
    def __init__(self, stdin):
        self.stdin = stdin
        self.buffer = []
        self.ptr = 0

    def next(self):
        if self.ptr >= len(self.buffer):
            self.buffer = self.stdin.readline().split()
            self.ptr = 0
        res = self.buffer[self.ptr]
        self.ptr += 1
        return res

    def nextInt(self):
        return int(self.next())

def main():
    in_ = FastScanner(sys.stdin)
    n = in_.nextInt()
    a = []
    for _ in range(n):
        line = in_.next()
        a.append(list(line))
    
    P = 37
    powP = [1] * (n + 1)
    for i in range(1, n + 1):
        powP[i] = powP[i - 1] * P
    
    hashRow = [[0] * n for _ in range(n)]
    hashCol = [[0] * n for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            c = a[i][j]
            val = ord(c)
            for k in range(n):
                exponent = (j - k + n) % n
                term = val * powP[exponent]
                hashRow[i][k] += term
                exponent_col = (i - k + n) % n
                term_col = val * powP[exponent_col]
                hashCol[j][k] += term_col
    
    ans = 0
    for A in range(n):
        for B in range(n):
            ok = True
            for i in range(n):
                row = (i + A) % n
                col = (i + B) % n
                val_row = hashRow[row][B]
                val_col = hashCol[col][A]
                if val_row != val_col:
                    ok = False
                    break
            if ok:
                ans += 1
    print(ans)

if __name__ == "__main__":
    main()
# End of Code
